### PR TITLE
Add constraint mergeability

### PR DESCRIPTION
**Pull request by jamie256**
## Description
Updates for constraints with some minor incidental bug fixes.
* Changed *Constraints to store constraints as dictionaries.
* Defined merge operations on constraints.
* Made default param handling consistent.
* Added some basic tests, need more in this area.

This should let us integrate with profile logging and not lose constraints after merging profiles. The merging is better when the constraints are accessible by name.

### General Checklist

* [x] Tests added for this feature/bug
      if it was a bug, test must cover it.
* [x] Conform by the style guides, by using formatter
* [ ] Documentation updated
* [ ] (optional) Please add a label to your PR

### CODE REVIEW

**Review by cswarth — APPROVED** [review]
LGTM

File: src/whylogs/core/statistics/constraints.py [hunk 1]
```diff
@@ -251,60 +279,104 @@ def report(self):
...
-    def __init__(self, constraints: List[ValueConstraint] = []):
-        self.constraints = constraints
+    def __init__(self, constraints: Mapping[str, ValueConstraint] = None):
+        if constraints is None:
+            constraints = dict()
+
+        if isinstance(constraints, list):
+            self.constraints = {constraint.name: constraint for constraint in constraints}
+        else:
+            self.constraints = constraints
```
Comment (richard-rogers) [inline src/whylogs/core/statistics/constraints.py:289]: Should this deep copy?

File: src/whylogs/core/statistics/constraints.py [hunk 2]
```diff
@@ -114,6 +114,17 @@ def update(self, v) -> bool:
             if self._verbose:
                 logger.info(f"value constraint {self.name} failed on value {v}")
 
+    def merge(self, other) -> "ValueConstraint":
```
Comment (richard-rogers) [inline src/whylogs/core/statistics/constraints.py:117]: I've never seen type annotations as literal strings before, but it looks like it's valid.

File: src/whylogs/core/statistics/constraints.py [hunk 3]
```diff
@@ -251,60 +279,104 @@ def report(self):
...
     def update(self, v):
-        for c in self.constraints:
+        for c in self.constraints.values():
             c.update(v)
 
+    def merge(self, other) -> "ValueConstraints":
+        if not other or not other.constraints:
+            return self
+
+        merged_constraints = other.constraints.copy()
```
Comment (richard-rogers) [inline src/whylogs/core/statistics/constraints.py:319]: deep copy?

**Pull request merged**